### PR TITLE
util: all file writing is now done atomically by renaming a temp file

### PR DESCRIPTION
**Pull request by orndorffgrant**
## Test Steps

All CI should still pass

## Desired commit type

 - [ ] Squash and merge: Using the "Proposed Commit Message"
 - [ ] Create a merge commit and use "Proposed Commit Message"
 - [x] Rebase and merge, no merge commit, rely only on existing commit messages

## Checklist

 - [ ] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

## Does this PR require extra reviews?

 - [ ] Yes
 - [x] No

### CODE REVIEW

File: uaclient/system.py [hunk 1]
```diff
@@ -370,12 +371,22 @@ def write_file(filename: str, content: str, mode: int = 0o644) -> None:
...
+        logging.debug(
+            "Writing file %s atomically via tempfile %s", filename, tmpf.name
+        )
+        tmpf.write(content.encode("utf-8"))
+        tmpf.flush()
+        tmpf.close()
+        os.chmod(tmpf.name, mode)
+        os.makedirs(os.path.dirname(filename), exist_ok=True)
+        os.rename(tmpf.name, filename)
+    except Exception as e:
```
Comment (lucasmoura) [inline uaclient/system.py:386]: Maybe this is an overkill, but should we add an unit test for this code ? Just for the case where the write part fails and we call the `flush` method.

But let me know what you think

**Review by lucasmoura — APPROVED** [review]

File: uaclient/system.py (hunk 1 above)
Comment (orndorffgrant) [reply uaclient/system.py:386]: Good idea, I'll add one

Comment (lucasmoura) [reply uaclient/system.py:386]: Yes, I think that covers it perfectly

**Comment by lucasmoura** [comment]
Merging it once CI is green

**Review by renanrodrigo — APPROVED** [review]
love it

**Pull request merged**